src/julia_to_gap.jl

```diff
@@ -93,148 +131,235 @@ function julia_to_gap(x::Rational{T}) where {T<:Integer}
+      else
+        ret_val[i] = GapObj_internal(obj[i, :], nothing, Val(recursive))
+      end
     end
     return ret_val
 end
 
 ## Tuples
-function julia_to_gap(
+function GapObj_internal(
     obj::Tuple,
-    recursion_dict::IdDict{Any,Any} = IdDict();
-    recursive::Bool = false,
-)
+    recursion_dict::GapCacheDict,
+    ::Val{recursive},
+) where recursive
     array = collect(Any, obj)
-    return julia_to_gap(array, recursion_dict; recursive)
+    return GapObj_internal(array, recursion_dict, Val(recursive))
 end
 
 ## Ranges
-function julia_to_gap(r::AbstractRange{<:Integer}, recursive::Bool = false)
+GAP.@install function GapObj(r::AbstractRange{<:Integer})
     res = NewRange(length(r), first(r), step(r))
     Wrappers.IsRange(res) || throw(ConversionError(r, GapObj))
     return res
 end
 
-function julia_to_gap(
-    r::AbstractRange{<:Integer},
-    recursion_dict::IdDict{Any,Any};
-    recursive::Bool = false)
-
-    return julia_to_gap(r)
-end
-
 ## Dictionaries
-function julia_to_gap(
+function GapObj_internal(
     obj::Dict{T,S},
-    recursion_dict::IdDict{Any,Any} = IdDict();
-    recursive::Bool = false,
-) where {S} where {T<:Union{Symbol,AbstractString}}
+    recursion_dict::GapCacheDict,
+    ::Val{recursive},
+) where {S} where {T<:Union{Symbol,AbstractString}} where {recursive}
+
+    recursive && recursion_dict !== nothing && haskey(recursion_dict, obj) && return recursion_dict[obj]
+
+    ret_val = NewPrecord(0)
 
-    record = NewPrecord(0)
-    if recursive
-        recursion_dict[obj] = record
+    rec = recursive && _needs_tracking_julia_to_gap(S)
+    if rec && recursion_dict === nothing
+        recursion_dict = RecDict()
     end
+
+    if recursion_dict !== nothing
+      recursion_dict[obj] = ret_val
+    end
+
     for (x, y) in obj
         x = Wrappers.RNamObj(MakeString(string(x)))
         if recursive
-            y = get!(recursion_dict, y) do
-                julia_to_gap(y, recursion_dict; recursive)
-            end
+          if rec
+            res = GapObj_internal(y, recursion_dict::RecDict, Val(recursive))
+          else
+            res = GapObj_internal(y, nothing, Val(recursive))
+          end
+        else
+            res = y
         end
-        Wrappers.ASS_REC(record, x, y)
+        Wrappers.ASS_REC(ret_val, x, res)
     end
 
-    return record
+    return ret_val
 end
 
 ## GAP objects:
 ## We have to do something only if recursive conversion is required,
 ## and if `obj` contains Julia subobjects;
 ## in this case, `obj` is a GAP list or record.
-## An example of such an `obj` is `GAP.julia_to_gap([[1]])`.
-function julia_to_gap(
+## An example of such an `obj` is `GapObj([[1]])`.
+function GapObj_internal(
     obj::GapObj,
-    recursion_dict::IdDict{Any,Any} = IdDict();
-    recursive::Bool = false,
-)
+    recursion_dict::GapCacheDict,
+    ::Val{recursive},
+) where {recursive}
     if ! recursive
         ret_val = obj
     elseif Wrappers.IsList(obj)
```

Comment:
This method maybe needs
```julia
recursion_dict !== nothing && haskey(recursion_dict, obj) && return recursion_dict[obj]
```

Of course this then is an existing bug.

It might matter if converting e.g. a `Vector{GapObj}` recursively which contains the same GAP object multiple times, and that object needs conversion...

On the other I am not even sure anymore whether this entire method is desirable. There is an argument for stopping conversion once we reach a `GapObj`. (Of course one could allow a third value for `recursive` to distinguish between "stop at GAP objects" and "recurse through everything", but I really wonder how useful it is to recurse through GapObj as we do here. Do we have any real world example where this ever mattered?

Reply:
First of all: Yes, this extra line is needed if we want recursive conversion of `GapObj`'s to GAP; I will add a test for this situation.

Concerning the question whether we want this recursive conversion (and analogously recursive conversion of Julia objects to Julia), I think it would be not a problem for users to change the current setup, and define that recursive conversion to GAP stops at `GapObj`, and recursive conversion to Julia stops at non-`GapObj`.
Probably just the tests in GAP.jl will notice the difference.

Reply:
We discussed this at the OSCAR workshop and I think we agreed that we can stop at GapObj, and that's it.

Can also be handled in a follow-up PR, though we should perhaps open an issue to remind us (unless you want to handle it immediately)